### PR TITLE
ET-1279: Changes to support AMP

### PR DESCRIPTION
**Pull request by anand-venkatraman**

### CODE REVIEW

**Comment by ProfessorEugene** [comment]
LGTM

File: src/main/adapters/headertag/headertag.js [hunk 1]
```diff
@@ -11,6 +11,8 @@ var PulsePointHeaderTag = function(config) {
     var timeout = config.timeout ? config.timeout : 500;
     var startTime = null;
     var slots = {};
+    var bidsAvailableCallback = config.bidsAvailable;
```
Comment (ProfessorEugene) [inline src/main/adapters/headertag/headertag.js]: consider renaming to ```doneCallback``` or something

File: src/main/adapters/headertag/headertag.js [hunk 2]
```diff
@@ -78,6 +87,75 @@ var PulsePointHeaderTag = function(config) {
...
+            bidsAvailableCallbackDone = true;
+        }
+    }
+
+    /**
+    * Registers message event. When the creative is rendered in a cross domain 2nd iframe (ex. in AMP Google Glade tag),
+    * it sends a message back to the parent frame requesting the creative, and this listener will return the creative.
+    */
+    function registerMessageEvent() {
+        window.addEventListener("message", function(event) {
```
Comment (ProfessorEugene) [inline src/main/adapters/headertag/headertag.js]: For the guy that raises these events (which you just explained to me offline is the DFP creative itself) - consider doing an iteration over all window parents instead of a "window.parent.postMessage" to make the code a little bit more resilient to future rendering changes.

File: src/test/adapters/headertag/PulsePointHeaderTagTest.test.js [hunk 3]
```diff
@@ -308,6 +308,101 @@ describe('HeaderTag tests', function(){
...
+                    bidUrl: window.bidderUrl + 'bidOrtb',
+                    slots: [{
+                        cp: 500015,
+                        ct: tagId,
+                        cf: '300X250',
+                        placement: placement,
+                        elementId: div
+                    }
+                ],
+                bidsAvailable: function(targeting) {
```
Comment (ProfessorEugene) [inline src/test/adapters/headertag/PulsePointHeaderTagTest.test.js]: This is a good test, but it would be nice to add a test that also tests the ```window.postMessage``` functionality.

File: src/main/adapters/headertag/headertag.js (hunk 1 above)
Comment (anand-venkatraman) [inline src/main/adapters/headertag/headertag.js]: done.

File: src/main/adapters/headertag/headertag.js (hunk 2 above)
Comment (anand-venkatraman) [inline src/main/adapters/headertag/headertag.js]: will do, will send a separate PR for the creative changes itself.

File: src/test/adapters/headertag/PulsePointHeaderTagTest.test.js (hunk 3 above)
Comment (anand-venkatraman) [inline src/test/adapters/headertag/PulsePointHeaderTagTest.test.js]: The next test verifies that.

File: src/main/adapters/headertag/creative.js [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    registerMessageEvents();
+  }
+
+  /**
+  * Registers the events to write the creative.
+  */
+  function registerMessageEvents() {
+    window.addEventListener("message", function( event ) {
+        var data = safeParseMessage(event.data);
+        if(data.source === 'pulsepoint' && data.creative) {
```
Comment (ProfessorEugene) [inline src/main/adapters/headertag/creative.js]: Will this work if there's multiple ads on one page - e.g. 3 DFP ads in 3 different crossdomain iframes?  Maybe we should also add a ```data.adId``` property and make sure we're listening for the appropriate creative to render?

File: src/main/adapters/headertag/creative.js [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+    var ifr = document.createElement("iframe");
+    ifr.width = width;
+    ifr.height = height;
+    ifr.style.border = "0px";
+    ifr.scrolling = "no";
+    ifr.frameborder = 0;
+    ifr.allowtransparency = !0;
+    document.body.appendChild(ifr);
+    var ifrDoc = (ifr.contentWindow) ? ifr.contentWindow.document : (ifr.contentDocument.document) ? ifr.contentDocument.document : ifr.contentDocument;
+    if(requiresDocOpen()) ifrDoc.open();
```
Comment (ProfessorEugene) [inline src/main/adapters/headertag/creative.js]: We should start to consider using the 'javascript:..' approach described [at the end of this post](https://sparecycles.wordpress.com/2012/03/08/inject-content-into-a-new-iframe/) everywhere.

File: src/main/adapters/headertag/creative.js (hunk 4 above)
Comment (anand-venkatraman) [inline src/main/adapters/headertag/creative.js]: The response ends up to the source window of the ```getCreative``` call, so this should be fine.

File: src/main/adapters/headertag/creative.js (hunk 5 above)
Comment (anand-venkatraman) [inline src/main/adapters/headertag/creative.js]: seems to be working fine, will use this.
